network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -0,0 +1,151 @@
+	authorizedRolesMap[cborcodec.CodeClusterBlockProposal] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeClusterBlockVote] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeClusterBlockResponse] = flow.RoleList{flow.RoleCollection}
+
+	// collections, guarantees & transactions
+	authorizedRolesMap[cborcodec.CodeCollectionGuarantee] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeTransactionBody] = flow.RoleList{flow.RoleCollection}
+	authorizedRolesMap[cborcodec.CodeTransaction] = flow.RoleList{flow.RoleCollection}
+
+	// core messages for execution & verification
+	authorizedRolesMap[cborcodec.CodeExecutionReceipt] = flow.RoleList{flow.RoleExecution}
+	authorizedRolesMap[cborcodec.CodeResultApproval] = flow.RoleList{flow.RoleVerification}
+
+	// execution state synchronization
+	// NOTE: these messages have been deprecated
+	authorizedRolesMap[cborcodec.CodeExecutionStateSyncRequest] = flow.RoleList{}
+	authorizedRolesMap[cborcodec.CodeExecutionStateDelta] = flow.RoleList{}
+
+	// data exchange for execution of blocks
+	authorizedRolesMap[cborcodec.CodeChunkDataRequest] = flow.RoleList{flow.RoleVerification}
+	authorizedRolesMap[cborcodec.CodeChunkDataResponse] = flow.RoleList{flow.RoleExecution}
+
+	// result approvals
+	authorizedRolesMap[cborcodec.CodeApprovalRequest] = flow.RoleList{flow.RoleConsensus}
+	authorizedRolesMap[cborcodec.CodeApprovalResponse] = flow.RoleList{flow.RoleVerification}
+
+	// generic entity exchange engines
+	authorizedRolesMap[cborcodec.CodeEntityRequest] = flow.RoleList{flow.RoleAccess, flow.RoleConsensus, flow.RoleCollection} // only staked access nodes
+	authorizedRolesMap[cborcodec.CodeEntityResponse] = flow.RoleList{flow.RoleCollection, flow.RoleExecution}
+
+	// testing
+	authorizedRolesMap[cborcodec.CodeEcho] = flow.Roles()
+
+	// dkg
+	authorizedRolesMap[cborcodec.CodeDKGMessage] = flow.RoleList{flow.RoleConsensus} // sn nodes for next epoch
+}
+
+// AuthorizedSenderValidator using the getIdentity func will check if the role of the sender
+// is part of the authorized roles list for the type of message being sent. A node is considered
+// to be authorized to send a message if all of the following are true.
+// 1. The message type is a known message type (initialized in the authorizedRolesMap).
+// 2. The authorized roles list for the message type contains the senders role.
+// 3. The node has a weight > 0 and is not ejected
+func AuthorizedSenderValidator(log zerolog.Logger, getIdentity func(peer.ID) (*flow.Identity, bool)) MessageValidator {
+	log = log.With().
+		Str("component", "authorized_sender_validator").
+		Logger()
+
+	return func(ctx context.Context, from peer.ID, msg *message.Message) pubsub.ValidationResult {
+		identity, ok := getIdentity(from)
+		if !ok {
+			log.Warn().Str("peer_id", from.String()).Msg("could not get identity of sender")
+			return pubsub.ValidationReject
+		}
+
+		msgType := msg.Payload[0]
```

Comment:
We could include a comment here as well noting the dependency on the encoding being used.

Reply:
Maybe we could add a method to the cbor codec module that returns this value to make the dependency explicit